threedi_schema_edits/breach_location.py

```diff
@@ -0,0 +1,484 @@
+
+from qgis.core import QgsFeatureRequest
+from qgis.core import QgsFeature
+from qgis.core import QgsGeometry
+from qgis.core import QgsVectorLayer
+from qgis.core import QgsPoint
+from qgis.core import QgsField
+
+from PyQt4.QtCore import QVariant
+
+from ThreeDiToolbox.utils.geo_utils import calculate_perpendicular_line
+from ThreeDiToolbox.utils.geo_utils import get_distance
+from ThreeDiToolbox.utils.geo_utils import get_epsg_code_from_layer
+from ThreeDiToolbox.utils.geo_utils import get_extrapolated_point
+
+from ThreeDiToolbox.utils.utils import parse_db_source_info
+
+from ThreeDiToolbox.threedi_schema_edits.predictions import Predictor
+from ThreeDiToolbox.utils import constants
+from ThreeDiToolbox.utils import utils
+
+log = logging.getLogger(__name__)
+
+EXTRAPLORATION_RATIO = 20
+
+
+class BreachLocation(object):
+
+    def __init__(self, search_distance, distance_to_levee, use_selection,
+                 is_dry_run, connected_pnt_lyr):
+        """
+        :param search_distance: the distance to search for a levee intersection
+        :param distance_to_levee: the distance from the levee to place the
+            point at
+        :param use_selection: use the selected features
+        :param is_dry_run: do not save the outcome of the tool to the DB yet.
+            Will produce two memory layers to visualize the results
+        :param connected_pnt_lyr: the connected point layer
+
+        """
+
+        self.search_distance = search_distance
+        self.distance_to_levee = distance_to_levee
+        self.use_selection = use_selection
+        self.is_dry_run = is_dry_run
+
+        self.selected_pnt_ids = collections.defaultdict(list)
+
+        self.connected_pnt_lyr = connected_pnt_lyr
+        self.fnames_conn_pnt = {
+            field.name(): i for i, field
+            in enumerate(self.connected_pnt_lyr.pendingFields())
+        }
+        if not self.is_dry_run:
+            self.connected_pnt_lyr.startEditing()
+        source_info = self.connected_pnt_lyr.dataProvider().dataSourceUri()
+        source_info_dict = parse_db_source_info(source_info)
+        predictor = Predictor(source_info_dict['db_type'])
+        uri = predictor.get_uri(**source_info_dict)
+        self.calc_pnt_lyr = predictor.get_layer_from_uri(
+            uri, constants.TABLE_NAME_CALC_PNT, 'the_geom')
+        self.levee_lyr = predictor.get_layer_from_uri(
+            uri, constants.TABLE_NAME_LEVEE, 'the_geom')
+
+        self.epsg_code = get_epsg_code_from_layer(self.connected_pnt_lyr)
+
+        if self.is_dry_run:
+            self.create_tmp_layers()
+        if self.use_selection:
+            self.set_selected_pnt_ids()
+
+        if self.epsg_code == constants.EPSG_WGS84:
+            self.search_distance /= constants.DEGREE_IN_METERS
+            self.distance_to_levee /= constants.DEGREE_IN_METERS
+
+        # used for user feedback
+        self.cnt_moved_pnts = 0
+
+    @property
+    def has_valid_selection(self):
```

Comment:
If I'm not mistaken, there is only one variable (self.selected_pnt_ids) that determines the return value of this method.